services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/multipart/MultipartDownloaderSubscriber.java

```diff
@@ -198,4 +201,15 @@ private GetObjectRequest nextRequest(int nextPartToGet) {
             }
         });
     }
+
+    private void validatePartsCount(int currentGetCount) {
+        if (totalParts != null && currentGetCount != totalParts) {
+            String errorMessage = "PartsCount validation failed. Expected " + totalParts + ", downloaded"
+                                  + " " + currentGetCount + " parts.";
+            log.error(() -> errorMessage);
```

Comment:
We should probably not log using error level if we are going to throw an exception anyway